src/Visualization/Python/GenerateXdmf.py

```diff
@@ -60,196 +60,246 @@ def generate_xdmf(file_prefix, output, subfile_name, start_time, stop_time,
+                    # In 2d we cannot read any z-coordinates because they
+                    # were never written.
+                    components.remove(coordinates + '_z')
+                components.remove('connectivity')
+                if is_2d_surface_in_3d_space:
+                    components.remove('pole_connectivity')
+                components.remove('total_extents')
+                components.remove('grid_names')
+                components.remove('bases')
+                components.remove('quadratures')
 
-        # close time grid
-        xdmf_output += "  </Grid>\n"
+                # Write the tensors that are to be visualized.
+                for component in components:
+                    if component.endswith("_x"):
+                        # Write a vector using the three components that make up
+                        # the vector (i.e. v_x, v_y, v_z)
+                        vector = component[:-2]
+                        xdmf_output += (
+                            "      <Attribute Name=\"%s\" "
+                            "AttributeType=\"Vector\" Center=\"Node\">\n" %
+                            (vector))
+                        xdmf_output += data_item_vec
+
+                        index_list = ["_x", "_y"]
+                        if dimensionality == 3:
+                            index_list.append("_z")
+                        for index in index_list:
+                            xdmf_output += (data_item(
+                                h5temporal.get(vector + index).dtype) +
+                                            Grid_path + "/%s" % (vector) +
+                                            index + "\n" +
+                                            "        </DataItem>\n")
+                        xdmf_output += "        </DataItem>\n"
+                        xdmf_output += "      </Attribute>\n"
+                    elif (component.endswith("_y")
+                          or component.endswith("_z")):
+                        # The component is a y or z component of a vector
+                        # it will be processed with the x component
+                        continue
+                    else:
+                        # If the component is not part of a vector,
+                        # write it as a scalar
+                        xdmf_output += (
+                            "      <Attribute Name=\"%s\" "
+                            "AttributeType=\"Scalar\" Center=\"Node\">\n" %
+                            (component))
+                        xdmf_output += data_item(
+                            h5temporal.get(component).dtype) + Grid_path + (
+                                "/%s\n" % component) + "        </DataItem>\n"
+                        xdmf_output += "      </Attribute>\n"
+                xdmf_output += "    </Grid>\n"
+
+            if is_2d_surface_in_3d_space:
+                if filling_poles_this_iteration:
+                    done = True
```

Comment:
I'm not sure if this is better, but you could also do `while True:` and then replace the `done = True` with a `break` statement. Up to you!

Reply:
I'm on the fence, and I'm inclined to break the degeneracy by leaving it as it is for expediency. But I don't see an obvious reason why one way would be better than the other

Reply:
Since we currently only have 2 types of volume data (surface and volume), I think the `done = True`  method is ok. In the future if we add more types of volume data or make the current types more complicated such that we'd have to go through this loop again, then the `while True: ... break` method would be better. So for now, I'd leave it.